internal/api/logs/v1/http.go

```diff
@@ -0,0 +1,149 @@
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	ReadTimeout        = 15 * time.Minute
+	WriteTimeout       = time.Minute
+	lokiUpstreamPrefix = "/loki"
+)
+
+type handlerConfiguration struct {
+	logger     log.Logger
+	registry   *prometheus.Registry
+	instrument handlerInstrumenter
+}
+
+// HandlerOption modifies the handler's configuration
+type HandlerOption func(h *handlerConfiguration)
+
+// Logger add a custom logger for the handler to use.
+func Logger(logger log.Logger) HandlerOption {
+	return func(h *handlerConfiguration) {
+		h.logger = logger
+	}
+}
+
+// Registry adds a custom Prometheus registry for the handler to use.
+func Registry(r *prometheus.Registry) HandlerOption {
+	return func(h *handlerConfiguration) {
+		h.registry = r
+	}
+}
+
+// HandlerInstrumenter adds a custom HTTP handler instrument middleware for the handler to use.
+func HandlerInstrumenter(instrumenter handlerInstrumenter) HandlerOption {
+	return func(h *handlerConfiguration) {
+		h.instrument = instrumenter
+	}
+}
+
+type handlerInstrumenter interface {
+	NewHandler(labels prometheus.Labels, handler http.Handler) http.HandlerFunc
+}
+
+type nopInstrumentHandler struct{}
+
+func (n nopInstrumentHandler) NewHandler(labels prometheus.Labels, handler http.Handler) http.HandlerFunc {
+	return handler.ServeHTTP
+}
+
+func NewHandler(read, write *url.URL, opts ...HandlerOption) http.Handler {
+	c := &handlerConfiguration{
+		logger:     log.NewNopLogger(),
+		registry:   prometheus.NewRegistry(),
+		instrument: nopInstrumentHandler{},
+	}
+	for _, o := range opts {
+		o(c)
+	}
+
+	r := chi.NewRouter()
+
+	if read != nil {
+		var proxyRead http.Handler
+		{
+			read.Path = fmt.Sprintf("%s%s", read.Path, lokiUpstreamPrefix)
+			middlewares := proxy.Middlewares(
+				proxy.MiddlewareSetUpstream(read),
+				proxy.MiddlewareLogger(c.logger),
+				proxy.MiddlewareMetrics(c.registry, prometheus.Labels{"proxy": "logsv1-read"}),
+			)
+
+			proxyRead = &httputil.ReverseProxy{
+				Director: middlewares,
+				ErrorLog: proxy.Logger(c.logger),
+				Transport: &http.Transport{
+					DialContext: (&net.Dialer{
+						Timeout: ReadTimeout,
+					}).DialContext,
+				},
+			}
+		}
+		r.Handle("/api/v1/query", c.instrument.NewHandler(
+			prometheus.Labels{"group": "logsv1", "handler": "query"},
+			proxyRead,
+		))
+		r.Handle("/api/v1/query_range", c.instrument.NewHandler(
+			prometheus.Labels{"group": "logsv1", "handler": "query_range"},
+			proxyRead,
+		))
+
+		var readyProxy http.Handler
+		{
+			middlewares := proxy.Middlewares(
+				proxy.MiddlewareSetUpstream(read),
+				proxy.MiddlewareLogger(c.logger),
+				proxy.MiddlewareMetrics(c.registry, prometheus.Labels{"proxy": "logsv1-ready"}),
+			)
+
+			readyProxy = &httputil.ReverseProxy{
+				Director: middlewares,
+			}
+		}
+		r.Mount("/ready", c.instrument.NewHandler(
+			prometheus.Labels{"group": "logsv1", "handler": "ready"},
+			readyProxy,
+		))
```

Comment:
Can you explain why we would need to expose the ready endpoint to a user of Observatorium? Isn't this an internal implementation detail?
Like, we might want to use that `/ready` endpoint to check with Kubernetes if the endpoint of a Pod is to be added to a Service, but not expose it to the user direct, am I missing something about Loki here? :)

Reply:
I introduced this initially to add [here](https://github.com/observatorium/observatorium/pull/47/files#diff-7ddfb3e035b42cd70649cc33393fe32cR217) an extra healthcheck ReadinessCheck agains loki for observatorium.